pkg/vpc/cleanup.go

```diff
@@ -30,49 +33,50 @@ func findAvailableNetworkInterfaces(provider api.ClusterProvider, spec *api.Clus
 			},
 		},
 	}
-	output, err := provider.EC2().DescribeNetworkInterfaces(input)
-	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
-	}
+
 	securityGroupRE, err := regexp.Compile(fmtSecurityGroupNameRegexForCluster(spec.Metadata.Name))
 	if err != nil {
-		return nil, errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
+		return errors.Wrap(err, "failed to create security group regex")
 	}
-	networkInterfaces := []string{}
-	for _, eni := range output.NetworkInterfaces {
-		id := *eni.NetworkInterfaceId
-		for _, sg := range eni.Groups {
-			if securityGroupRE.MatchString(*sg.GroupName) {
-				logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
-				networkInterfaces = append(networkInterfaces, id)
-				break
-			} else {
+
+	var lastErr error
+
+	err = ec2API.DescribeNetworkInterfacesPagesWithContext(ctx, input, func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) bool {
+		for _, eni := range output.NetworkInterfaces {
+			id := *eni.NetworkInterfaceId
+			for _, sg := range eni.Groups {
+				if securityGroupRE.MatchString(*sg.GroupName) {
+					logger.Debug("found %q, which belongs to our security group %q (%s)", id, *sg.GroupName, *sg.GroupId)
+					if err := fn(id); err != nil {
+						lastErr = err
+						return false
+					}
+					break
+				}
 				logger.Debug("found %q, but it belongs to security group %q (%s), which does not appear to be ours", id, *sg.GroupName, *sg.GroupId)
-				break
 			}
 		}
+		return !lastPage
+	})
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to list dangling network interfaces in %q", spec.VPC.ID)
 	}
-	return networkInterfaces, nil
+
+	return lastErr
 }
 
-func deleteNetworkInterfaces(provider api.ClusterProvider, networkInterfaces []string) error {
-	for _, eni := range networkInterfaces {
+// CleanupNetworkInterfaces finds and deletes any dangling ENIs
+func CleanupNetworkInterfaces(ec2API ec2iface.EC2API, spec *api.ClusterConfig) error {
+	ctx := context.TODO()
```

Comment:
I would pass a context to `CleanupNetworkInterfaces` instead, or, if you are certain we can wait indefinitely (which I doubt) use `context.Background()` instead of `context.TODO()`

Reply:
The idea was to have `CleanupNetworkInterfaces` eventually take a context when it's available in the calling function. Since a context isn't available in the calling function, not sure it'd make sense to move this there.
`context.TODO()` is used here to be caught by a linter (if a rule is configured) and as a reminder to supply a meaningful context when it's available.

Reply:
I would propose to fix it in the calling function now. If it is too much work, please create an issue and fix it soon after this PR.

Reply:
If by fixing it, you mean making the `context.TODO()` or `context.Background()` call from the calling function and passing it, then it's a minor change. If we actually want a useful context to be passed in from the top-level caller and supply that to other function calls that support it (which is what I meant from my comment), that'd require more changes.

I can do the former in this PR but I don't understand the reasoning behind it.

Reply:
The reasoning behind it is not to make the situation worse than it was. I am happy to drop it if the overall context issue is fixed right after this PR.

Reply:
I think we should address all context issues in a separate PR as it's more involved and revert the changes here to use the non-context function calls. I'll create a separate issue for reviewing usage of context. WDYT?

Reply:
Yes, please just use whatever the default context SDK uses, like we have in other instances:
```
pkg/cfn/manager/api.go:	if err := c.provider.CloudFormation().ListStacksPages(input, pager); err != nil {
pkg/cfn/manager/api.go:	if err := c.provider.CloudFormation().DescribeStackEventsPages(input, pager); err != nil {
pkg/cfn/manager/api.go:	if err := c.provider.CloudTrail().LookupEventsPages(input, pager); err != nil {
```